Radius

HelenMamontova opened the pull request:
Implemented so that the mod_radius works as a server in radius.

HelenMamontova committed:
Radius. Header file <mutex>, <jthread.hpp> added. Functions start, stop
declaration changed. Class members m_thread, m_mutex added. Function Run
declaration added.

HelenMamontova committed:
Radius. Header files added. Functions Start, Stop, Run definition added.

HelenMamontova committed:
Radius. New files server.h, server.cpp added.

HelenMamontova committed:
Radius. Definition of functions startReceive, makeResponse, handleSend,
handleReceive added.

HelenMamontova committed:
The Run function name fixed. Definition of secret variable sdded.

HelenMamontova committed:
The file server.cpp added to add_library command, command
find_package(OpenSSL 1.0.0 Required) added, OpenSSL::Crypto added to
target_link_libraries command in the block if(BUILD_MOD_RADIUS).

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -17,3 +23,34 @@ RADIUS::RADIUS()
+int RADIUS::Stop()
+{
+    std::lock_guard lock(m_mutex);
+    isRunning = false;
```
`isRunning` must be set to `false` in the `Run` method, before exiting.

Also, see the `ping` plugin and look how `stop_token` and `m_thread.request_stop()` are used. In particular, notice that we join the thread in `Stop` after requesting stop (or detaching it in case we failed to stop the thread function).

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -17,3 +23,34 @@ RADIUS::RADIUS()
+int RADIUS::Run()
+{
+    std::string secret("secret");
+    uint16_t port = 1812;
```
Pass those values directly.

HelenMamontova committed:
Parameter token added to function Run() declaration.

HelenMamontova committed:
Radius. Method Start: variable isRunning=true removed, m_thread
definition changed. Method Stop: variable isRunning=false removed,
m_thread.joinable check added, isRunning check added, request_stop call added.
Method Run: parameter token added, variables secret, port removed,
object lock and isRunning=true added before cycle while,
cycle while added, isRunning=false added after cycle while.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -17,3 +23,48 @@ RADIUS::RADIUS()
+
+    m_thread.request_stop();
+
+
```
Extra empty line.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -17,3 +23,48 @@ RADIUS::RADIUS()
+
+int RADIUS::Stop()
+{
+    std::lock_guard lock(m_mutex);
```
You do not need to hold the mutex here.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -17,3 +23,48 @@ RADIUS::RADIUS()
+
+
+    if (isRunning)
+        m_thread.detach();
```
Let's just go for `join` for now. Unconditionally.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -17,3 +23,48 @@ RADIUS::RADIUS()
+
+int RADIUS::Run(std::stop_token token)
+{
+    std::lock_guard lock(m_mutex);
```
This lock has a function-level scope. It will be released when the function finishes. It means that no one can access `isRunning` while the function is running. You should hold locks only when you read or write a variable, `isRunning` in particular, and release them as soon as possible - after reading or writing.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -17,3 +23,48 @@ RADIUS::RADIUS()
+    std::lock_guard lock(m_mutex);
+    isRunning = true;
+
+    while (!token.stop_requested())
```
This will not work. `io_service.run()` will run indefinitely and you won't reach the second iteration of this cycle. Let's just skip checking the `token` fro now.

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -24,7 +26,10 @@ namespace STG
             mutable std::string errorStr;
+            std::jthread m_thread;
+            std::mutex m_mutex;
             bool isRunning;
```
I know that the old plugins use `isRunning` and `errorStr`, but in the new code let's stick to the `m_` naming scheme for the private members. So please change them to `m_running` and `m_errorStr`.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -17,3 +23,48 @@ RADIUS::RADIUS()
+    {
+        try
+        {
+            boost::asio::io_service io_service;
```
Let's call it `ioService` or `ios`. We do not use underscore in variable and function names.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -17,3 +23,48 @@ RADIUS::RADIUS()
+        }
+        catch (const std::exception& e)
+        {
+            std::cerr << "Exception: " << e.what() <<"\n";
```
Avoid using `std::cout` and `std::cerr`. See how the other plugins report errors.

madf commented on projects/stargazer/plugins/other/radius/server.h:
```diff
@@ -0,0 +1,23 @@
+#include <optional>
+#include <cstdint> //uint8_t, uint32_t
+
+class Server
```
Please put it into the `STG` namespace.

HelenMamontova committed:
Radius. Hold the mutex removed,extra conditions
for m_thread.join() removed in the Stop function.

HelenMamontova committed:
Radius. Cycle while removed in function Run.

HelenMamontova committed:
Radius. The variables isRunning, errorStr replaced by m_running,
m_errorStr in class Radius.

HelenMamontova committed:
Radius. The variable isRunning replaced by m_running in function Run.

HelenMamontova committed:
Radius. Object name io_service changed to ioService in the function Run.

HelenMamontova committed:
Method SetRunning declaration added to the class RADIUS.

HelenMamontova committed:
Method SetRunning definition added. Method SetRunning call added to
function Run.

HelenMamontova committed:
Radius. Namespace STG added.

HelenMamontova committed:
Radius. Declaration using STG::Server added.

HelenMamontova committed:
Radius. Header file "stg/logger.h" added. RADIUS class member m_loger added.

HelenMamontova committed:
Radius. Header file "stg/common.h" added. Initialization of m_logger
added to constructor RADIUS. Output cerr replaced by logger and
printfd().

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -14,6 +23,48 @@ std::string RADIUS::GetVersion() const
+    }
+    catch (const std::exception& e)
+    {
+        m_errorStr = "Exceptoin by Run()";
```
'Exception' -> 'Exception' (misprint).

'by Run()' -> 'in RADIUS::Run'

Also please add `e.what()`.

Same below.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -14,6 +23,48 @@ std::string RADIUS::GetVersion() const
+int RADIUS::Stop()
 {
+    if (m_thread.joinable())
+        m_thread.join();
```
You must join a thread only after you have requested it to stop. Because before requesting, the join would hang forever, because the thread does not know it is expected to stop. And because the join hangs, you can not request the thread to stop.

The logic here should be as follows:
 - if the thread is not joinable, nothing to be done here, just return;
 - otherwise, request stop;
 - join the thread to wait until it stops.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -14,6 +23,48 @@ std::string RADIUS::GetVersion() const
+
+void RADIUS::SetRunning(bool val)
+{
+    std::lock_guard lock(m_mutex);
```
All `std::lock_guard` instances can and should be `const`.

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -10,21 +14,26 @@ namespace STG
-            bool isRunning;
+            mutable std::string m_errorStr;
+            std::jthread m_thread;
+            std::mutex m_mutex;
```
A little pro-tip. Always put the mutex first in the member list. This way you ensure that it is initialized when you use it.

I had an issue in the past when I put the mutex after the thread, like you here. And I started the thread right inside the initialization, and it ran before the mutex was initialized, and inside the thread function I tried to lock the mutex and everything crashed, because the mutex was not initialized. It was a really tricky issue to find, because most time the thread starts with a little delay, and the mutex has a chance to be initialized.

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -10,21 +14,26 @@ namespace STG
+            bool m_running;
 
+            int Run(std::stop_token token);
+            PluginLogger m_logger;
```
Please put the `m_logger` after the `m_running`, in the member list.

HelenMamontova committed:
Radius. Variable except added, messages fixed in the function Run.

HelenMamontova committed:
Radius. Thread join logic fixed in the function Stop.

HelenMamontova committed:
Keyword const added to std::lock_guard in the function SetRunning.

HelenMamontova committed:
Radius. Class member m_mutex moved to the top of list.

HelenMamontova committed:
Radius. Class member m_logger put after the m_running.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -60,9 +62,10 @@ int RADIUS::Run(std::stop_token token)
-        m_logger("Exception by Run: %s", e.what());
-        printfd(__FILE__, "Exception by Run. Message: '%s'\n", e.what());
+        std::string except(e.what());
+        m_errorStr = "Exception in RADIUS::Run(): " + except;
```
Consider `+ std::string(e.what());`

HelenMamontova committed:
Radius. Unnecessary variable except removed in the function Run.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -14,6 +23,50 @@ std::string RADIUS::GetVersion() const
+
+    m_thread.request_stop();
+
+    m_thread. join();
```
A whitespace after '.'

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -10,21 +14,26 @@ namespace STG
+            int Stop() override;
             int Reload(const ModuleSettings & /*ms*/) override { return 0; }
-            bool IsRunning() override { return isRunning; }
+            bool IsRunning() override { return m_running; }
```
Returning `m_running` must be guarded by mutex.

HelenMamontova committed:
Radius. Extra whitespace removed in the function Stop. Function
IsRunning definition added.

HelenMamontova committed:
Radius. Function IsRunning declaration changed.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -50,6 +50,12 @@ void RADIUS::SetRunning(bool val)
 }
 
+bool RADIUS::IsRunning()
+    {
```
Bad indentation.

HelenMamontova committed:
Radius. Formatting fixed.

madf commented on projects/stargazer/plugins/other/radius/server.h:
```diff
@@ -0,0 +1,26 @@
+#include <boost/asio.hpp>
+#include <optional>
+#include <cstdint> //uint8_t, uint32_t
+                   //
```
Unused `//`

HelenMamontova committed:
Radius. Extra symbols '//' removed.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -14,6 +23,56 @@ std::string RADIUS::GetVersion() const
 }
 
 RADIUS::RADIUS()
+    : m_logger(PluginLogger::get("radius"))
```
`m_running` must be also initialized here.

HelenMamontova committed:
Radius. Class member m_running initialization added to constructor
RADIUS.
